ext/soap/php_wss.c

```diff
@@ -0,0 +1,279 @@
+			break;
+		case SOAP_WSS_DIGEST_METHOD_SHA512:
+			algo = zend_string_init("sha512", strlen("sha512"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA512));			
+			break;
+	}
+
+	if (ops == NULL) {
+		if (algo != NULL) {
+			php_error_docref(NULL, E_WARNING, "invalid hashing algorithm, %s", ZSTR_VAL(algo));
+			zend_string_release_ex(algo, 0);
+		} else {
+			php_error_docref(NULL, E_WARNING, "unknown hashing algorithm");
+		}
+		return 0;
+	}
+
+	zend_string_release_ex(algo, 0);
+
+	zend_string *exc_c14n = get_exc_c14n_from_node_ptr(source);
+	if (exc_c14n != NULL) {
+		void *context = php_hash_alloc_context(ops);
+		ops->hash_init(context, NULL);
+		ops->hash_update(context, (unsigned char *)ZSTR_VAL(exc_c14n), ZSTR_LEN(exc_c14n));
+		
+		zend_string *digest = zend_string_alloc(ops->digest_size, 0);
+		ops->hash_final((unsigned char *)ZSTR_VAL(digest), context);
+		efree(context);
+
+		ZSTR_VAL(digest)[ops->digest_size] = 0;
+
+		zend_string *digest_base64 = php_base64_encode((unsigned char *)ZSTR_VAL(digest), ZSTR_LEN(digest));
+		xmlNewChild(reference, ns_ds, BAD_CAST("DigestValue"), BAD_CAST(ZSTR_VAL(digest_base64)));
+
+		zend_string_release(digest_base64);
+		zend_string_release(digest);
+		zend_string_release_ex(exc_c14n, 0);
+
+		return 1;
+	} else {
+		php_error_docref(NULL, E_WARNING, "exclusive C14N failed, URI \"%s\"", ZSTR_VAL(source_uri));
+		return 0;
+	}
+}
+/* }}} */
+
+void add_wss_to_function_call(zval *this_ptr, int soap_version, xmlNodePtr envelope, xmlNodePtr head, xmlNodePtr body) /* {{{ */
+{
+	char random_s[10];
+	if (Z_TYPE_P(Z_CLIENT_WSS_RANDOM_ID_P(this_ptr)) == IS_TRUE) {
+		slprintf(random_s, sizeof(random_s), "%d", (int)php_mt_rand_common(100000000,999999999));
+	} else {
+		slprintf(random_s, sizeof(random_s), "%d", 1);
+	}
+
+	zend_string *body_id = zend_string_concat2(
+		"BodyID-", strlen("BodyID-"),
+		random_s, strlen(random_s));
+	zend_string *body_id_uri = zend_string_concat2(
+		"#", strlen("#"),
+		ZSTR_VAL(body_id), ZSTR_LEN(body_id));
+	
+	xmlNsPtr ns_wsse = NULL;
+	if (Z_LVAL_P(Z_CLIENT_WSS_VERSION_P(this_ptr)) == SOAP_WSS_VERSION_1_1) {
+		ns_wsse = xmlNewNs(envelope, BAD_CAST(SOAP_WSS_SECEXT_1_1), BAD_CAST("wsse11"));
+	} else {
+		ns_wsse = xmlNewNs(envelope, BAD_CAST(SOAP_WSS_SECEXT_1_0), BAD_CAST("wsse"));
+	}
+
+	xmlNsPtr ns_ds = xmlNewNs(envelope, BAD_CAST(SOAP_WSS_XMLDSIG), BAD_CAST("ds"));
+	xmlNsPtr ns_wsu = xmlNewNs(envelope, BAD_CAST(SOAP_WSS_UTILITY), BAD_CAST("wsu"));
+	xmlNodePtr security = xmlNewChild(head, ns_wsse, BAD_CAST("Security"), NULL);
+	xmlSetNsProp(body, ns_wsu, BAD_CAST("Id"), BAD_CAST(ZSTR_VAL(body_id)));
+
+	if (soap_version == SOAP_1_1) {
+		xmlSetProp(security, BAD_CAST(SOAP_1_1_ENV_NS_PREFIX":mustUnderstand"), BAD_CAST("1"));
+	} else {
+		xmlSetProp(security, BAD_CAST(SOAP_1_2_ENV_NS_PREFIX":mustUnderstand"), BAD_CAST("true"));
```

Comment:
Can any of the `xml*` calls fail due to out of memory?

Reply:
Possibly if the **libxml** handles everything in memory, but then the same goes for the rest of the SOAP extension where `xml*` calls are used.